core/src/ModelArray.cpp

```diff
@@ -0,0 +1,332 @@
+/*!
+ * @file ModelData.cpp
+ *
+ * @date Feb 24, 2022
+ * @author Tim Spain <timothy.spain@nersc.no>
+ */
+
+#include "include/ModelArray.hpp"
+
+#include <algorithm>
+#include <cstdarg>
+#include <iterator>
+#include <set>
+#include <string>
+#include <utility>
+
+namespace Nextsim {
+
+ModelArray::SizeMap ModelArray::m_sz;
+ModelArray::DimensionMap ModelArray::m_dims;
+
+const std::map<ModelArray::Type, std::string> ModelArray::typeNames = {
+    { ModelArray::Type::H, "HField" },
+    { ModelArray::Type::U, "UField" },
+    { ModelArray::Type::V, "VField" },
+    { ModelArray::Type::Z, "ZField" },
+    { ModelArray::Type::DG, "DGHField--DO-NOT-USE--" },
+};
+
+ModelArray::ModelArray(const Type type, const std::string& name)
+    : type(type)
+    , m_name(name)
+{
+    m_data.resize(std::max(std::size_t { 0 }, m_sz.at(type)), nComponents());
+}
+
+ModelArray::ModelArray()
+    : ModelArray(Type::H, "")
+{
+}
+
+ModelArray::ModelArray(const ModelArray& orig)
+    : ModelArray(orig.type, orig.m_name)
+{
+    setData(orig.m_data);
+}
+
+ModelArray& ModelArray::operator=(const ModelArray& orig)
+{
+    type = orig.type;
+    m_name = orig.m_name;
+    setData(orig.m_data);
+
+    return *this;
+}
+
+ModelArray& ModelArray::operator=(const double& fill)
+{
+    m_name = std::to_string(fill);
+    setData(fill);
+
+    return *this;
+}
+
```

Comment:
Hi Tim, I noticed all the math operators return new ModelArray instances. I was wondering if it would be useful to add in-place operators as well.

Reply:
Done

Reply:
Hi Tim, by in-place I meant to add non-const operators that modify the ModelArray's internal data in-place and do not create a new ModelArray instance. This might be useful as well.